Lower-level Nexus access utilities

SimonHeybrock opened the pull request:
This adds `nexus.Group`, which can used as follows:

```python
from scippneutron import nexus
from scippneutron.nexus import NX_class
f = nexus.File(filename)  # should use contextmanager, but that is cumbersome in a Jupyter notebook

f.by_nx_class()[NX_class.NXlog]['SampleTemp']  # nexus.Group (wrapping h5py.Group), nothing loaded yet
f.by_nx_class()[NX_class.NXlog]['SampleTemp']['time', :3]
f.by_nx_class()[NX_class.NXlog]['proton_charge'][:1000]  # dim label redundant if single dim
f.by_nx_class()[NX_class.NXmonitor]['monitor1'][...]  # load everything
f.by_nx_class()[NX_class.NXevent_data]['bank42_events']['pulse', :20000]
```

The implementation right now is a draft. I am considering the following changes:

- `nexus.Group` implementation has cumbersome branching bases on `NX_class`. My current idea would be to turn this into a base-class, `NXobject`, which can be implemented, e.g., by `NXlog`, `NXmonitor`, and `NXevent_data`.
- The child classes provide properties such as `dims, `shape`, and `unit`.
- Child classes provide `__getitem__` to actually load the class. This would forward to the current internal functions for loading such groups.

Does it make sense to introduce such an inheritance-based design? Or is there a better option?

jl-wynen commented on src/scippneutron/nexus/nexus.py:
```diff
@@ -0,0 +1,91 @@
+        self._loader = LoadFromHdf5()
+
+    def __getattr__(self, name):
+        return getattr(self._group, name)
```
Seems a bit odd. What is this used for?

SimonHeybrock replied:
To forward to methods of `h5py.Group`, for things we are not implementing ourselves. Might not be what we want for the final product, but useful while implementation is still incomplete.

jl-wynen commented on src/scippneutron/nexus/nexus.py:
```diff
@@ -0,0 +1,91 @@
+    NXevent_data = auto()
+
+
+class Group:
```
Putting this here so we have a dedicated thread.

> nexus.Group implementation has cumbersome branching bases on NX_class

We need this branching somewhere. Are you saying it is cumbersome because it needs to be repeated in several methods?

> Child classes provide __getitem__ to actually load the class. This would forward to the current internal functions for loading such groups.

So `NXobject.__getitem__` would return an instance one of those child classes if the underlying HDF5 object is not a group? I think this makes sense.

> Does it make sense to introduce such an inheritance-based design? Or is there a better option?

It would certainly make sense for establishing a common interface. At least until we can use https://docs.python.org/3/library/typing.html#typing.Protocol (Introduced in Python 3.8. Can't be long until we drop support for 3.7...)
But what else would the base class contain? Default implementations of `dims` etc. that raise an exception? I m not sure there is much value in using inheritance here unless there are methods which actually do something that can be reused by the child classes.

SimonHeybrock replied:
> We need this branching somewhere. Are you saying it is cumbersome because it needs to be repeated in several methods?

Yes.

> So NXobject.__getitem__ would return an instance one of those child classes if the underlying HDF5 object is not a group? I think this makes sense.

Rather: Would return an instance of the HDF5 object is a group with known `NX_class` attribute. The `NXobject` class map to HDF5 groups, not datasets. But yes otherwise.

> But what else would the base class contain? Default implementations of dims etc. that raise an exception? I m not sure there is much value in using inheritance here unless there are methods which actually do something that can be reused by the child classes.

I think there will likely be a number of common methods, e.g., simply for tree traversal. Basically everything that is in `h5py.Group` would be in the base class. Will see while implementing if, e.g., composition makes more sense.

jl-wynen replied:
> Rather: Would return an instance of the HDF5 object is a group with known NX_class attribute. The NXobject class map to HDF5 groups, not datasets. But yes otherwise.

I don't understand. Where do the derived classes come in?

SimonHeybrock replied:
As you say: You access a group entry, if it is an `h5py.Group` (not a `h5py.Dataset`) then it should usually have a `NX_class` attribute, and instead of returning the base class `NXobject` (wrapping `h5py.Group`) we return, e.g., an `NXlog` instance if `NX_class == 'NXlog'`.

jl-wynen replied:
Ok. Then I misunderstood your earlier reply.

I would say that inheritance is a good solution at first order. Like you said, we should see what the base class would actually be like.

Tom-Willemsen replied:
I think I agree, the inheritance-based solution sounds ok to me, but I think we need to see what the code would actually look like to make more detailed comments...

Tom-Willemsen commented:
Perhaps I've not understood some aspect of the API, but why do we want:

```
f.by_nx_class()[NX_class.NXlog]['SampleTemp']
```

rather than the more natural looking

```
f.by_nx_class(NX_class.NXlog)['SampleTemp']
```

?

jl-wynen commented:
We could also 'overload' `__getitem__` and use
```.py
f[NX_class.NXlog]['SampleTemp']
```
Since the argument of the first brackets is of type `NX_class`, we could identify what the user means and call `by_nx_class` internally.
@SimonHeybrock, this would correspond to your initial suggestion, I think. Why did you move away from it?

SimonHeybrock commented:
> We could also 'overload' `__getitem__` and use
> 
> ```python
> f[NX_class.NXlog]['SampleTemp']
> ```
> 
> Since the argument of the first brackets is of type `NX_class`, we could identify what the user means and call `by_nx_class` internally. @SimonHeybrock, this would correspond to your initial suggestion, I think. Why did you move away from it?



> We could also 'overload' `__getitem__` and use
> 
> ```python
> f[NX_class.NXlog]['SampleTemp']
> ```
> 
> Since the argument of the first brackets is of type `NX_class`, we could identify what the user means and call `by_nx_class` internally. @SimonHeybrock, this would correspond to your initial suggestion, I think. Why did you move away from it?

`by_nx_class` does a recursive walk of the tree, it felt odd that `__getitem__` would operate recursively.

SimonHeybrock commented:
@jl-wynen @Tom-Willemsen Please have a look at the new `nxobject.py`, and subclasses of `NXobject` that I added. Essentially some of the lower-level functions become implementation details of the classes (and they could each and individually be refactored to do this in a nicer way than the current plain wrapper).

SimonHeybrock commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,74 @@
+
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        cls._registry[cls.__name__] = cls
```
> to the nexus class names from a data file. 

What data file are you referring to?

jl-wynen replied:
`make` uses the content of the 'NX_class' attribute to look up the subclass which is registered by its name. So the name of the subclass is coupled to how the nexus class is stored in the file.

SimonHeybrock replied:
Well, not the file, but by which `NX_class` values the NeXus standard defines? Not sure I understand the comments, can you elaborate?

jl-wynen replied:
The names of the Python classes must be the same as the `NX_class` names in NeXus. We could decouple the two by using separately constructed strings to look up the registry. E.g. `cls._registry[cls.NX_class] = cls` (assuming a class property `NX_class`

SimonHeybrock replied:
> The names of the Python classes must be the same as the NX_class names in NeXus.

Yes, isn't that good?

> We could decouple the two by using separately constructed strings to look up the registry

We could, but why? Isn't that adding confusion?

SimonHeybrock commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,74 @@
+    @property
+    def NX_class(self):
+        nx_class = self._loader.get_string_attribute(self._group, 'NX_class')
+        return NX_class[nx_class]
```
I had considered getting rid of the enum, using the actual classes as dict keys. That works, but when printing them, the class names are displayed in full, including their "path" within the package and submodules, which was hard to read. If I moved them, e.g., to `scippneutron.nexus` it would be reasonable, maybe? Or do you know of a way to change the `__repr__` of a type?

jl-wynen replied:
> Or do you know of a way to change the __repr__ of a type?

This would make it hard to analyse and debug some code. And I don't think it is possible. At least not without a custom metaclass.

jl-wynen commented on src/scippneutron/file_loading/_common.py:
```diff
@@ -76,3 +76,38 @@ def _add_attr_to_loaded_data(attr_name: str,
+    elif isinstance(select, int) or isinstance(select, slice):
+        check_1d()
+        return select
+    elif isinstance(select, dict):
+        index = [slice(None)] * len(dims)
+        for key, sel in select.items():
+            if not ignore_missing and key not in dims:
+                raise ValueError(
+                    f"'{key}' used for indexing not found in dataset dims {dims}.")
+            index[dims.index(key)] = sel
+        return tuple(index)
```
We don't support indexing by dict elsewhere. I think we should either not do it here either or support it in Variable and DataArray.

SimonHeybrock replied:
We have to support it here, since we cannot use `group['x', 4]['y', 5]`, since the first `__getitem__` would already load the entire column.

SimonHeybrock replied:
But yes, the plan is to support this in scipp sooner or later.

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,74 @@
+            if hasattr(item, 'visititems'):
+                return self.make(item)
+            else:
+                return item
```
Does this return `h5.Dataset` objects? I would have expected that we wrap them.

SimonHeybrock replied:
Well, Nexus does not really define much for "fields" (their name for HDF5 datasets), so such a class would have limited use, apart from hiding differences between and underlying HDF5 vs. Json, which is not the main aim of this PR. We can't do much with an individual dataset (it cannot be loaded as a scipp object). Many NXobject subclasses map to `sc.DataArray`, but Nexus fields have insufficient information to load them, e.g., as a `sc.Variable`.

But a wrapper could certainly be added later if required?

jl-wynen replied:
Why even support loading it at all if it is not really useful?

SimonHeybrock replied:
I am not saying that it is not useful, just that there is nothing/little Nexus-specific about it, so users could fall back to plain h5py.

jl-wynen replied:
Than I would still wrap it in order to generalise over HDF5 and JSON. A user can then use a property to get the underlying `h5.Dataset` or json dict. This would make it more explicit what is happening.

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,74 @@
+
+    @functools.lru_cache()
+    def by_nx_class(self):
+        classes = self._loader.find_by_nx_class(tuple(self._registry), self._group)
```
This behaves differently depending on what modules were imported because that can influence the registry. We have to make sure that we always import all modules containing subclasses of `NX_object`.
That is, this approach seems fragile.

SimonHeybrock replied:
If it is not imported in cannot be created either? Not sure how that could be done differently.

jl-wynen replied:
We could create the registry in a specific file so that it is always fully available if that file is imported. But this would likely mean moving the subclass construction logic out of the base class. Which I am in favour of anyway.

SimonHeybrock replied:
Doesn't this lead to some cyclic import issue?

jl-wynen replied:
We could have `nx_registry.py` with `registry=None` and in `__init__.py` import the various `NX_xxx` classes and do `nx_registry.registry = [...]`. `NX_object` could then use `nx_registry.py` to instantiate a concrete object because that module never imports anything itself.

SimonHeybrock replied:
@jl-wynen I switched to a manual registry, is this along the lines of what you wanted?

jl-wynen replied:
Not quite. This has a circular dependency between the modules. It only avoids problems by delaying the import. I was thinking of moving the registry into a module for `NX_object` to depend on. (Or leaving it in the same module, that part is not so important.) And populating the registry in a place that depends on `NX_object`, `NX_log`, etc. This way, there are no circular dependencies between modules. A bit like your original solution but with a centralised code for populating the registry.

SimonHeybrock replied:
Sounds more complicated. Is it really that much better?

jl-wynen replied:
Not sure. I think it would be cleanest because it has only one way dependencies between modules. But the current solution should work just fine as long as no one calls `_nx_class_registry` during import.

SimonHeybrock commented:
I am encountering some weirdness with the existing implementation:

- `group.name` is used in some cases, which works only if `group` is either a `h5py.Group` or a `JSONGroup`.
  - Some operations with `LoadFromJson` do not appear to return a `JSONGroup`, so we end up with an exception (previously untested code path?). I replaced this with a `get_path` method, which for now just returns the current name in `LoadFromJson`. In this particular case it only affects a log message.
- `find_by_nx_class` included the current group for `LoadFromHdf5` but not in `LoadFromJson`. I changed the former, tests pass, but I don't know if I broke loading of some files?

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,132 @@
+        self._dataset = dataset
+        self._loader = loader
+
+    def __getitem__(self, index):
```
It would be good to have more type annotations.
(Putting the comment here because this is the first place where I noticed it. There are a bunch more.)

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,132 @@
+    NXevent_data = auto()
+
+
+class Field:
```
In `h5py`, groups and datasets have a ref to their file. This can be useful for navigation and comprehensive error messages. I think it would be a good idea to have something similar in all nexus objects.

SimonHeybrock replied:
Have a ref to their file? Isn't that in the group, i.e., not need to duplicate?

jl-wynen replied:
How would a user access it, though?

SimonHeybrock replied:
We would add a `file` property, just like `h5py.Group` has one? Not sure I understand the question.

jl-wynen replied:
I was referring to

> Isn't that in the group, i.e., not need to duplicate?

The information is present but not accessible by the user.

SimonHeybrock replied:
The interface is *very* incomplete right now, there are potentially dozens of others. I suggest we add things when we need them.

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,132 @@
+
+    @property
+    def unit(self):
+        # TODO Prefer to return None if no such attr, provided that scipp supports this
```
Add a test to make sure we don't forget when scipp is updated?

SimonHeybrock replied:
I changed it directly now, to return `None`.

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,132 @@
+    NXevent_data = auto()
+
+
+class Field:
```
Is `Field` not a subtype of `NXobject` in NeXus? That is, are `NXobjects` always HDF5 groups?

SimonHeybrock replied:
Yes, `NXobject` is always a group.

jl-wynen commented on src/scippneutron/file_loading/nxobject.py:
```diff
@@ -0,0 +1,132 @@
+            out[NX_class[nx_class]] = {n: self._make(g) for n, g in zip(names, groups)}
+        return out
+
+    @property
+    def nx_class(self):
+        key = self._loader.get_string_attribute(self._group, 'NX_class')
+        return NX_class[key]
```
How about turning this into a class attribute? It would mean all derived classes have to define it, but there no need to load it dynamically, right?

SimonHeybrock replied:
Erm, not sure: Thinks such as `NXmonitor` and `NXdetector` "kind of" inherit `NXdata` (though there is no actualy inheritance defined). I am just working on the former to create `NXdata` (or `NXevent_data`) to deal with concrete loading. Not sure if we want `nx_class` to refer the actual attribute from the file, or the current instance being used to process the group?

jl-wynen replied:
Are you saying that there will be, .e.g, an `NXdata` object in python processing an `NXmonitor` group in HDF5?
It seems odd to then use `nx_class == NXmonitor` in this case because, using the `NXdata` object, the monitor-specific contents are not accessible. But I guess, I am not sure what `nx_class` is used for.

SimonHeybrock replied:
We could probably just remove the `nx_class` property. I have added an `attrs` property in a follow up branch, so the user could just do `group.attrs['NX_class']` instead of `group.nx_class`. Should I remove this convenience property?

jl-wynen replied:
Yes, I think that makes sense.

SimonHeybrock replied:
Ran into an "issue" right away when trying to remove: Most files to not correctly set `NX_class` for the root (to `NXroot`), so the convenience property does actually add value here (avoiding special case handling in client code). I suggest we keep it?

jl-wynen replied:
Ok. But then add a docstring, please.

SimonHeybrock replied:
done in follow-up

jl-wynen commented on src/scippneutron/file_loading/_detector_data.py:
```diff
@@ -199,19 +202,88 @@ def _load_detector(group: Group, nexus: LoadFromNexus) -> DetectorData:
+    @property
+    def unit(self):
+        # Binned data, bins do not have a unit
+        return sc.units.one
```
Can we already return `None` here?

SimonHeybrock replied:
Done in follow-up

jl-wynen commented on src/scippneutron/file_loading/_detector_data.py:
```diff
@@ -199,19 +202,88 @@ def _load_detector(group: Group, nexus: LoadFromNexus) -> DetectorData:
+        data = detector_data.event_data
+        data.bins.coords['id'] = data.bins.coords.pop('detector_id')
+        data.bins.coords['time_offset'] = data.bins.coords.pop('tof')
+        data.coords['time_zero'] = data.coords.pop('pulse_time')
```
What are these names based on?

SimonHeybrock replied:
Names from `NXevent_data`, without the `event_` prefix. Not sure this is a good choice, could keep the full name?

jl-wynen replied:
I think it makes the most sense if these low level functions load the data as faithfully as possible.

SimonHeybrock replied:
Done in follow-up

jl-wynen commented on src/scippneutron/file_loading/_detector_data.py:
```diff
@@ -199,19 +202,88 @@ def _load_detector(group: Group, nexus: LoadFromNexus) -> DetectorData:
+                      select=...) -> DetectorData:
     _check_for_missing_fields(group, nexus)
+    index = to_plain_index([_pulse_dimension], select)
+    if isinstance(index, tuple):
```
`index` can be an empty tuple. I think that should be equivalent to an ellipsis and not raise `IndexError`. Maybe `to_plain_index` should handle it in order to not complicate the code here.

SimonHeybrock replied:
Done in follow-up

jl-wynen commented on src/scippneutron/file_loading/_detector_data.py:
```diff
@@ -199,19 +202,88 @@ def _load_detector(group: Group, nexus: LoadFromNexus) -> DetectorData:
+    def shape(name):
+        return nexus.get_shape(nexus.get_dataset_from_group(group, name))
+
+    max_index = shape("event_index")[0]
+    single = False
+    if index is Ellipsis:
+        last_loaded = False
+    else:
+        if isinstance(index, int):
+            single = True
+            start, stop, _ = slice(index, None).indices(max_index)
+            if start == stop:
+                raise IndexError('')
+            index = slice(start, start + 1)
+        start, stop, stride = index.indices(max_index)
+        if stop + stride > max_index:
+            last_loaded = False
+        else:
+            stop += stride
+            last_loaded = True
+        index = slice(start, stop, stride)
```
This seems very complicated. Could it be simplified if `to_plain_index` returned a slice instead of int or ellipsis?
Also, it does not handle the case when index is a dict.

SimonHeybrock replied:
I don't think `to_plain_index` can simplify much here. The problem is that we need to deal with loading one more index (if possible), since `NXevent_data` does not define "end" indices. This is different from general slicing applications.

jl-wynen commented on src/scippneutron/file_loading/_detector_data.py:
```diff
@@ -199,19 +202,88 @@ def _load_detector(group: Group, nexus: LoadFromNexus) -> DetectorData:
+            single = True
+            start, stop, _ = slice(index, None).indices(max_index)
+            if start == stop:
+                raise IndexError('')
```
Message?

SimonHeybrock replied:
Fixed in follow-up

jl-wynen commented on src/scippneutron/file_loading/_detector_data.py:
```diff
@@ -230,32 +302,42 @@ def _load_event_group(group: Group, nexus: LoadFromNexus, detector_data: Detecto
-    ends = sc.concat(
-        [event_index[_pulse_dimension, 1:],
-         sc.scalar(number_of_event_ids)], _pulse_dimension)
+    begins = event_index[_pulse_dimension, :-1]
+    ends = event_index[_pulse_dimension, 1:]
```
Put into `else` block?

SimonHeybrock replied:
Done in follow up branch.

SimonHeybrock commented on src/scippneutron/nexus/nexus.py:
```diff
@@ -0,0 +1,17 @@
+
+class File(AbstractContextManager, NXroot):
+    def __init__(self, *args, **kwargs):
+        self._file = h5py.File(*args, **kwargs)
```
Not sure. How would you do it?

jl-wynen replied:
```.py
def __enter__(self) -> File:
    self._file.__enter__()
    return self
```

SimonHeybrock replied:
But we are not using the `h5py.File` as context manager? What would calling `__enter__` achieve?

jl-wynen replied:
Then why does it inherit from `AbstractContextManager`?

SimonHeybrock replied:
Because it is implementing a context manager? But it is not using the `h5py.File` context manager. But maybe that is possible, I just don't know how?

jl-wynen replied:
>> Then why does it inherit from AbstractContextManager?
>
> Because it is implementing a context manager?

I misread your comment, I though you said, we are not using `nexus.File` as a context manager.
In this case, we should properly enter and exist `h5py.File`, even though `h5py.File.__enter__` is probably a noop, just like for `nexus.File` right now.

My suggestion for `__enter__` above would do that. Also, `__exit__` should exit `self._file` instead of `self._group`. And in order to let users use `nexus.File` outside of context managers, we can also provide
```.py
def close():
    self._file.close()
```
This way, we are really wrapping the `h5py.File`.

SimonHeybrock replied:
Done, seems to work.

jl-wynen replied:
On the other branch?

SimonHeybrock replied:
Ah, yes, sorry.

SimonHeybrock commented on tests/test_nexus.py:
```diff
@@ -0,0 +1,196 @@
+        classes = root.by_nx_class()
+        assert list(classes[nexus.NX_class.NXentry]) == ['entry']
+        assert list(classes[nexus.NX_class.NXmonitor]) == ['monitor']
+        assert list(classes[nexus.NX_class.NXlog]) == ['log']
```
It can be used to load something with a know specific name, without having to manually search the entire tree structure. Someone may be interested in a specific log, but not know in which group and subgroup it is stored.

jl-wynen replied:
Ah, I checked the implementation again. It returns a dict of dicts. The test made me think it returns a dict of lists.

SimonHeybrock commented on src/scippneutron/file_loading/_hdf5_nexus.py:
```diff
@@ -148,19 +153,24 @@ def load_dataset(self,
+            return variable
+        return sc.array(dims=dimensions,
+                        unit=self.get_unit(dataset),
+                        values=dataset[index].astype(dtype))
```
Would be nice, can't think of a solution right now. Maybe there is a helper in numpy? I am adding a TODO note for now, unless you have a better idea?

SimonHeybrock commented on src/scippneutron/file_loading/_monitor_data.py:
```diff
@@ -1,42 +1,76 @@
 
+def load_monitor(group: Group, nexus: LoadFromNexus, select=tuple()) -> sc.DataArray:
+    # Look for event mode data structures in NXMonitor. Event-mode data takes
+    # precedence over histogram-mode-data if available.
```
done in follow-up

SimonHeybrock commented on tests/test_nexus.py:
```diff
@@ -0,0 +1,196 @@
+    with resource(builder_with_events_monitor_and_log())() as f:
+        root = nexus.NXroot(f, loader)
+        assert root.nx_class == nexus.NX_class.NXroot
+        assert set(root.keys()) == set(['entry', 'monitor'])
```
done in follow-up

SimonHeybrock commented on tests/test_nexus.py:
```diff
@@ -0,0 +1,196 @@
+        assert root['/entry/log'].name == '/entry/log'
+
+
+def test_nxobject_by_nx_class_contains_everything(nexus_group: Tuple[Callable,
```
done in follow-up

SimonHeybrock commented on tests/test_nexus.py:
```diff
@@ -0,0 +1,196 @@
+    return builder
+
+
+def test_nxobject_tree_traversal(nexus_group: Tuple[Callable, LoadFromNexus]):
```
done in follow-up

jl-wynen reviewed (approved):
Assuming the next PR will address my comments in this one.
